block/internal/executing/executor.go

```diff
@@ -0,0 +1,644 @@
+			Data:  res.BatchData,
+		}, common.ErrNoTransactionsInBatch
+	}
+
+	return &BatchData{
+		Batch: res.Batch,
+		Time:  res.Timestamp,
+		Data:  res.BatchData,
+	}, nil
+}
+
+// createBlock creates a new block from the given batch
+func (e *Executor) createBlock(ctx context.Context, height uint64, batchData *BatchData) (*types.SignedHeader, *types.Data, error) {
+	currentState := e.GetLastState()
+
+	// Get last block info
+	var lastHeaderHash types.Hash
+	var lastDataHash types.Hash
+
+	if height > e.genesis.InitialHeight {
+		lastHeader, lastData, err := e.store.GetBlockData(ctx, height-1)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get last block: %w", err)
+		}
+		lastHeaderHash = lastHeader.Hash()
+		lastDataHash = lastData.Hash()
+	}
+
+	// Get signer info
+	pubKey, err := e.signer.GetPublic()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get public key: %w", err)
+	}
+
+	// Build validator hash
+	// Get validator hash
+	validatorHash, err := e.options.ValidatorHasherProvider(e.genesis.ProposerAddress, pubKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get validator hash: %w", err)
+	}
+
+	// Create header
+	header := &types.SignedHeader{
+		Header: types.Header{
+			Version: types.Version{
+				Block: currentState.Version.Block,
+				App:   currentState.Version.App,
+			},
+			BaseHeader: types.BaseHeader{
+				ChainID: e.genesis.ChainID,
+				Height:  height,
+				Time:    uint64(batchData.UnixNano()),
+			},
+			LastHeaderHash:  lastHeaderHash,
+			ConsensusHash:   make(types.Hash, 32),
+			AppHash:         currentState.AppHash,
+			ProposerAddress: e.genesis.ProposerAddress,
+			ValidatorHash:   validatorHash,
+		},
+		Signer: types.Signer{
+			PubKey:  pubKey,
+			Address: e.genesis.ProposerAddress,
+		},
+	}
+
+	// Create data
+	data := &types.Data{
+		Txs: make(types.Txs, len(batchData.Transactions)),
+		Metadata: &types.Metadata{
+			ChainID:      header.ChainID(),
+			Height:       header.Height(),
+			Time:         header.BaseHeader.Time,
+			LastDataHash: lastDataHash,
+		},
+	}
+
+	for i, tx := range batchData.Transactions {
+		data.Txs[i] = types.Tx(tx)
+	}
+
+	// Set data hash
+	if len(data.Txs) == 0 {
+		header.DataHash = common.DataHashForEmptyTxs
+	} else {
+		header.DataHash = data.DACommitment()
+	}
+
+	return header, data, nil
+}
+
+// applyBlock applies the block to get the new state
+func (e *Executor) applyBlock(ctx context.Context, header types.Header, data *types.Data) (types.State, error) {
+	currentState := e.GetLastState()
+
+	// Prepare transactions
+	rawTxs := make([][]byte, len(data.Txs))
+	for i, tx := range data.Txs {
+		rawTxs[i] = []byte(tx)
+	}
+
+	// Execute transactions
+	ctx = context.WithValue(ctx, types.HeaderContextKey, header)
+	newAppHash, _, err := e.exec.ExecuteTxs(ctx, rawTxs, header.Height(),
+		header.Time(), currentState.AppHash)
+	if err != nil {
+		e.sendCriticalError(fmt.Errorf("failed to execute transactions: %w", err))
```

Comment:
A critical error is sent to the channel but also returned. The caller in the executionLoop just logs the error. 
This workflow can be confusing and may lead to concurrent operations. Rule of thump is: either handle the error or return it. In this case I would return it here and do the escalation in the execution loop

Reply:
Fair point. When I look at `produceBlock` errors, I am wondering if it is a good idea to ignore most of them. Maybe it is better to revisit the method and fail for every error returned. 
This can be done in a new PR . Not blocking this